### PR TITLE
Add support for LoRA and 4-bit quantization

### PR DESCRIPTION
**Pull request by orangetin**
Adds `--quantize` and `--lora-base BASE_MODEL_NAME` args

### CODE REVIEW

**Comment by orangetin** [comment]
Turning to draft because this implementation can be streamlined a little

**Comment by orangetin** [comment]
Ready for review @csris

**Comment by orangetin** [comment]
Example usage: `python3 serving_local_nlp_model.py --hf_model_name timdettmers/guanaco-7b --lora-base huggyllama/llama-7b --quantize`

**Comment by orangetin** [comment]
Actually, wait. Don't merge. I'm seeing some errors locally

**Comment by orangetin** [comment]
The model breaks when quantization is enabled with LoRA. We can run the model in fp16+lora and it should be fine. @csris how should I proceed?

**Comment by orangetin** [comment]
Ok, nvm. Fixed it by install transformers, accelerate, and peft from source

**Comment by orangetin** [comment]
Pending work item: Sync transformers_port

**Comment by orangetin** [comment]
Done. Tested the important bits locally.

File: dockerfile_together [hunk 1]
```diff
@@ -30,6 +30,8 @@ RUN pip3 install --no-cache-dir \
       sentencepiece \
       together_worker \
       uvicorn \
+      git+https://github.com/huggingface/peft.git \
```
Comment (csris) [inline dockerfile_together:33]: Relying on the HEAD revision of both `accelerate` and `peft` is a little risky. I'd suggest adding an `@GIT-HASH` to the end to pin to a specific commit. Not blocking this PR, though.

File: model_utils.py [hunk 2]
```diff
@@ -32,6 +35,18 @@ def get_local_huggingface_tokenizer_model(
...
         device_map = None
 
+    if quantize:
+        load_in_4bit = True
+        quantization_config = BitsAndBytesConfig(
+                load_in_4bit=True,
+                bnb_4bit_compute_dtype=torch.bfloat16,
+                bnb_4bit_use_double_quant=True,
+                bnb_4bit_quant_type='nf4'
+            )
```
Comment (csris) [inline model_utils.py:45]: Small formatting suggestion:
```suggestion
        quantization_config = BitsAndBytesConfig(
            load_in_4bit=True,
            bnb_4bit_compute_dtype=torch.bfloat16,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_quant_type='nf4'
        )
```

File: model_utils.py [hunk 3]
```diff
@@ -143,18 +158,45 @@ def get_local_huggingface_tokenizer_model(
             assert False
     elif model_path is not None and model_path != "":
         logger.warning("model_path is not None, but model_name is not given. Load from model_path only")
-        tokenizer = AutoTokenizer.from_pretrained(model_path, use_auth_token=auth_token, trust_remote_code=trust_remote_code)
-        model = AutoModelForCausalLM.from_pretrained(model_path, torch_dtype=torch.float16, use_auth_token=auth_token, device_map=device_map, trust_remote_code=trust_remote_code)
+        tokenizer = AutoTokenizer.from_pretrained(
+            model_path,
+            use_auth_token=auth_token,
+            trust_remote_code=trust_remote_code
+            )
```
Comment (csris) [inline model_utils.py:165]:
```suggestion
        tokenizer = AutoTokenizer.from_pretrained(
            model_path,
            use_auth_token=auth_token,
            trust_remote_code=trust_remote_code
        )
```

File: model_utils.py [hunk 4]
```diff
@@ -143,18 +158,45 @@ def get_local_huggingface_tokenizer_model(
...
+        model = AutoModelForCausalLM.from_pretrained(
+            model_path,
+            torch_dtype=torch.float16,
+            use_auth_token=auth_token,
+            device_map=device_map,
+            trust_remote_code=trust_remote_code
+            )
```
Comment (csris) [inline model_utils.py:172]:
```suggestion
        model = AutoModelForCausalLM.from_pretrained(
            model_path,
            torch_dtype=torch.float16,
            use_auth_token=auth_token,
            device_map=device_map,
            trust_remote_code=trust_remote_code
        )
```

File: model_utils.py [hunk 5]
```diff
@@ -143,18 +158,45 @@ def get_local_huggingface_tokenizer_model(
...
     else:
-        tokenizer = AutoTokenizer.from_pretrained(model_name, use_auth_token=auth_token, torch_dtype=dtype, trust_remote_code=trust_remote_code)
-        model = AutoModelForCausalLM.from_pretrained(model_name, use_auth_token=auth_token, device_map=device_map, trust_remote_code=trust_remote_code, torch_dtype=dtype)
+        tokenizer = AutoTokenizer.from_pretrained(
+            model_name,
+            use_auth_token=auth_token,
+            torch_dtype=dtype,
+            trust_remote_code=trust_remote_code
+            )
```
Comment (csris) [inline model_utils.py:179]:
```suggestion
        tokenizer = AutoTokenizer.from_pretrained(
            model_name,
            use_auth_token=auth_token,
            torch_dtype=dtype,
            trust_remote_code=trust_remote_code
        )
```

File: model_utils.py [hunk 6]
```diff
@@ -143,18 +158,45 @@ def get_local_huggingface_tokenizer_model(
...
+        model = AutoModelForCausalLM.from_pretrained(
+            model_name,
+            use_auth_token=auth_token,
+            device_map=device_map if lora_adapters != "" else None,
+            trust_remote_code=trust_remote_code,
+            torch_dtype=dtype,
+            load_in_4bit=load_in_4bit,
+            quantization_config=quantization_config,
+            )
```
Comment (csris) [inline model_utils.py:188]:
```suggestion
        model = AutoModelForCausalLM.from_pretrained(
            model_name,
            use_auth_token=auth_token,
            device_map=device_map if lora_adapters != "" else None,
            trust_remote_code=trust_remote_code,
            torch_dtype=dtype,
            load_in_4bit=load_in_4bit,
            quantization_config=quantization_config,
        )
```

**Review by csris — APPROVED** [review]
Left some comments, but none are blocking. Approved.

File: dockerfile_together (hunk 1 above)
Comment (orangetin) [reply dockerfile_together:33]: Makes sense, I'll make this change

File: model_utils.py (hunk 2 above)
Comment (orangetin) [reply model_utils.py:45]: thanks

**Pull request merged**